keps/sig-cluster-lifecycle/kubeadm/20190506-kubeadm-machine-output.md

```diff
@@ -0,0 +1,370 @@
+
+For more clarificaton, please see [this example](https://play.golang.org/p/_CJLB7gdLZQ) that highlights how buffered printer output versus unbuffered printer output might behave in the context of this KEP.
+
+##### Parity with kubectl and versioned output
+
+Parity with kubectl is defined as support for all of the output formats currently available to kubectl's `-o|--output` flag:
+
+  * `json`
+  * `yaml`
+  * `name`
+  * `template`
+  * `go-template`
+  * `go-template-file`
+  * `templatefile`
+  * `jsonpath`
+  * `jsonpath-file`
+
+The natural path to such parity would be to use the same mechanism in kubeadm as used by kubectl, the API machinery package `k8s.io/cli-runtime/pkg/printers`. However, the printers require input objects of type `runtime.Object`.
+
+Creating new or converting existing kubeadm objects to Kubernetes API-style objects has the immediate effect of introducing versioned output to kubeadm. It's the shared opinion of the authors of this KEP that versioned output should be a non-goal. This does not indicate an opinion on the value of versioned output, but rather acknowledges that such a design decision requires a much broader discussion.
+
+##### jq
+The program [`jq`](https://stedolan.github.io/jq/) is a performant, command-line solution for parsing and manipulating JSON.
+
+##### A friendly bootstrap token struct
+
+```golang
+struct {
+	ID          string    `json:"id"`
+	TTL         string    `json:"ttl"` // parseable by time.ParseDuration
+	Expires     time.Time `json:"expires"`
+	Usages      []string  `json:"usages"`
+	Description string    `json:"description"`
+}
+```
+
+##### Go template functions
+
+The Go template into which objects are emitted will have a function map that includes the following functions:
+
+| Signature | Description |
+|-----------|-------------|
+| `node0() string` | Returns the `addr:port` of the first node in a cluster |
+| `caCrt() string` | Returns the `sha256:hash` of the discovery token's CA certificate |
+| `join(list []string, sep string) string` | Calls `strings.Join(list, sep)` |
+
+##### Kubeadm init JSON output
+
+The following JSON output is an example of running `kubeadm init -o json`:
+
+```json
+{
+  "node0": "192.168.20.51:443",
+  "caCrt": "sha256:1f40ff4bd1b854fb4a5cf5d2f38267a5ce5f89e34d34b0f62bf335d74eef91a3",
+  "token": {
+    "id":          "5ndzuu.ngie1sxkgielfpb1",
+    "ttl":         "23h",
+    "expires":     "2019-05-08T18:58:07Z",
+    "usages":      [
+      "authentication",
+      "signing"
+    ],
+    "description": "The default bootstrap token generated by 'kubeadm init'.",
+    "extraGroups": [
+      "system:bootstrappers:kubeadm:default-node-token"
+    ]
+  },
+  "raw": "Rm9yIHRoZSBhY3R1YWwgb3V0cHV0IG9mIHRoZSAia3ViZWFkbSBpbml0IiBjb21tYW5kLCBwbGVhc2Ugc2VlIGh0dHBzOi8vZ2lzdC5naXRodWIuY29tL2FrdXR6LzdhNjg2ZGU1N2JmNDMzZjkyZjcxYjZmYjc3ZDRkOWJhI2ZpbGUta3ViZWFkbS1pbml0LW91dHB1dC1sb2c="
```

Comment:
I'd lean against the raw unless there is a concrete use case.

Reply:
Hi @timothysc,

We're going to remove `init` from the list of supported commands anyway as we'll be introducing the ability to reconstruct the original `kubeadm join` via a Go template and `token list`